Handle toUnicode cMaps that omit leading zeros in hex encoded UTF-16 (issue 18099)

alexcat3 opened the pull request:
Modifies partialEvaluator.readToUnicode() to handle toUnicode cMaps that omit leading zeros in hex encoded UTF-16, as seen in the PDF in [https://github.com/mozilla/pdf.js/issues/18099](issue 18099). 

In the PDF in question, the toUnicode cmap had a line to map the glyph char codes from 00 to 7F to the corresponding unicode code points. The syntax to map a range of char codes to a range of unicode code points is 
<start_char_code> <end_char_code> <start_unicode_codepoint>
As the unicode code points are supposed to be given in UTF 16, the PDF's line SHOULD have probably read
<00> <7F> <0000>
Instead it omitted two leading zeros from the UTF 16 like this
<00> <7F> <00>
This confused pdf.js into mapping these character codes to the UTF-16 characters with the corresponding high bytes ( 01 becomes \u0100, 02 becomes \u0200) which ended up turning latin text in the PDF into chinese when it was copied.

I'm not sure if the specification actually allows PDFs to do this, but since there's at least one PDF in the wild that does and other PDF readers read it correctly, pdf.js should probably support this.

calixteman commented:
Could you add a unit test ? something similar to:
https://github.com/mozilla/pdf.js/blob/790470c1aadf4dc6c4e8ff00364b87ecea4d7bd7/test/unit/api_spec.js#L3202-L3216
You can just test that the retrieved text starts with `Yýsledková listina přijímacích zkoušek`.

alexcat3 commented:
OK, I will.

Snuffleupagus commented on test/pdfs/issue18099_reduced.pdf:
It seems that this PDF document has a broken XRef table, given that opening it logs: `Warning: Indexing all PDF objects`.
Unless you specifically want to test handling of *corrupt* PDF documents, it's always best if they're valid. So, please run the file through a tool such as e.g. https://github.com/qpdf/qpdf since that should help clean-up the file.

alexcat3 replied:
Oops, sorry, i'll fix that

alexcat3 replied:
The pdf was created by hand by me, so that's how it got corrupted.

Snuffleupagus replied:
Also, please list the filename in https://github.com/mozilla/pdf.js/blob/master/test/pdfs/.gitignore

alexcat3 replied:
Done

Snuffleupagus commented on test/unit/api_spec.js:
```diff
@@ -3418,6 +3418,18 @@ Caron Broadcasting, Inc., an Ohio corporation (“Lessee”).`)
 
       await loadingTask.destroy();
     });
+    it("gets text content, correctly handling documents with toUnicode cmaps that omit leading zeros on hex-encoded UTF-16", async function () {
```
Please place a blank-line between test-cases to aid readability.
```suggestion
    
    it("gets text content, correctly handling documents with toUnicode cmaps that omit leading zeros on hex-encoded UTF-16", async function () {
```

timvandermeij replied:
Please keep a newline between tests, so in this case above `it(...)`, similar to the ones below.

Snuffleupagus commented on test/unit/api_spec.js:
```diff
@@ -3418,6 +3418,18 @@ Caron Broadcasting, Inc., an Ohio corporation (“Lessee”).`)
+        disableNormalization: true,
+      });
+      const text = mergeText(items);
+      expect(text).toEqual("Hello world!");
+    });
```
All unit-tests should clean-up after themselves, please note the other tests in this file.
```suggestion
      expect(text).toEqual("Hello world!");
      
      await loadingTask.destroy();
    });
```

Snuffleupagus reviewed (commented):
When addressing the comments, please remember to [squash the commits](https://github.com/mozilla/pdf.js/wiki/Squashing-Commits).

timvandermeij reviewed (commented):
Please squash the commits into one, see https://github.com/mozilla/pdf.js/wiki/Squashing-Commits. After that this looks good to me, but I'd like to ask @Snuffleupagus or @calixteman to give the final sign-off on this given their familiarity with the font code. Thanks!

alexcat3 committed:
Handle toUnicode cmaps that omit leading zeros in hex encoded UTF-16 (issue 18099)
Add unit test to check compatability with such cmaps

In the PDF in issue 18099. the toUnicode cmap had a line to map the glyph char codes from 00 to 7F to the corresponding code points. The syntax to map a range of char codes to a range of unicode code points is
<start_char_code> <end_char_code> <start_unicode_codepoint>
As the unicode code points are supposed to be given in UTF-16 BE, the PDF's line SHOULD have probably read
<00> <7F> <0000>
Instead it omitted two leading zeros from the UTF-16 like this
<00> <7F> <00>
This confused PDF.js into mapping these character codes to the UTF-16 characters with the corresponding HIGH bytes (01 became \u0100, 02 became \u0200, et cetera), which ended up turning latin text in the PDF into chinese when it was copied
I'm not sure if the PDF spec actually allows PDFs to do this, but since there's at least one PDF in the wild that does and other PDF readers read it correctly, PDF.js should probably support this

alexcat3 commented:
OK, I made the necessary changes and squashed the commits.

timvandermeij commented:
The CI failed on linting. You can automatically fix that by running `npx gulp lint --fix` locally. Other than that this patch LGTM, with passing tests which we'll trigger once the linting issue is fixed, but I'll leave the final sign-off here to @Snuffleupagus given the familiarity with the font code. Thanks!

calixteman commented:
/botio test

Snuffleupagus reviewed (changes requested):
Can you please improve the *commit message* a bit, such that all relevant information is available on the Git command line as well without having to read GitHub?

Generally the first line of the commit message should contain a summary of the changes, with a reference to the bug/issue, and then any other relevant details below. You included a bunch of nice context in https://github.com/mozilla/pdf.js/pull/18390#issue-2391419810 that belongs in the commit message too. My suggestion would be something like this:
```
Handle toUnicode cMaps that omit leading zeros in hex encoded UTF-16 (issue 18099)

In the PDF in question, the toUnicode cmap had a line to map the glyph char codes from 00 to 7F to the corresponding unicode code points. The syntax to map a range of char codes to a range of unicode code points is
<start_char_code> <end_char_code> <start_unicode_codepoint>
As the unicode code points are supposed to be given in UTF 16, the PDF's line SHOULD have probably read
<00> <7F> <0000>
Instead it omitted two leading zeros from the UTF 16 like this
<00> <7F> <00>
This confused PDF.js into mapping these character codes to the UTF-16 characters with the corresponding high bytes ( 01 becomes \u0100, 02 becomes \u0200) which ended up turning latin text in the PDF into chinese when it was copied.

I'm not sure if the specification actually allows PDFs to do this, but since there's at least one PDF in the wild that does and other PDF readers read it correctly, PDF.js should probably support this.
```

alexcat3 commented:
I changed the commit message

Snuffleupagus reviewed (approved):
r=me, thank you!

alexcat3 commented:
Thank YOU for all your help!
